syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt

```diff
@@ -14,145 +15,175 @@
+
+                            blockTempIdByHash[block.hash] = tempRepository.pushTempData(blockContent)
+
+                            updateProgress(blockContent.size.toLong())
+                        }
                     }
                 }
-                return progress()
-            }
 
-            override fun close() {
-                missingHashes.clear()
-                hashList.clear()
-                blocksByHash.clear()
-            }
-        }
-        synchronized(lock) {
-            hashList.addAll(fileBlocks.blocks.map { it.hash })
-            missingHashes.addAll(hashList)
-            for (block in fileBlocks.blocks) {
-                if (missingHashes.contains(block.hash)) {
-                    val requestId = Math.abs(Random().nextInt())
-                    requestIds.add(requestId)
-                    connectionHandler.sendMessage(Request.newBuilder()
-                            .setId(requestId)
-                            .setFolder(fileBlocks.folder)
-                            .setName(fileBlocks.path)
-                            .setOffset(block.offset)
-                            .setSize(block.size)
-                            .setHash(ByteString.copyFrom(Hex.decode(block.hash)))
-                            .build())
-                    logger.debug("sent request for block, hash = {}", block.hash)
+                fileBlocks.blocks.distinctBy { it.hash }.forEach { block ->
+                    pipe.send(block)
                 }
-            }
-            return fileDownloadObserver
-        }
-    }
 
-    fun onResponseMessageReceived(response: BlockExchangeProtos.Response) {
-        synchronized(lock) {
-            if (!requestIds.contains(response.id)) {
-                return
+                pipe.close()
             }
-            NetworkUtils.assertProtocol(response.code == ErrorCode.NO_ERROR, {"received error response, code = ${response.code}"})
-            val data = response.data.toByteArray()
-            val hash = Hex.toHexString(MessageDigest.getInstance("SHA-256").digest(data))
-            if (missingHashes.remove(hash)) {
-                blocksByHash.put(hash, data)
-                logger.debug("aquired block, hash = {}", hash)
-                lock.notify()
-            } else {
-                logger.warn("received not-needed block, hash = {}", hash)
+
+            // the sequence is evaluated lazy -> only one block per time is loaded
+            val fileBlocksIterator = fileBlocks.blocks
+                    .asSequence()
+                    .map { tempRepository.popTempData(blockTempIdByHash[it.hash]!!) }
+                    .map { ByteArrayInputStream(it) }
+                    .iterator()
+
+            return object : SequenceInputStream(object : Enumeration<InputStream> {
+                override fun hasMoreElements() = fileBlocksIterator.hasNext()
+                override fun nextElement() = fileBlocksIterator.next()
+            }) {
```

Comment:
How does this work? It seems you might be touching tempRepository before the block is downloaded?

Reply:
Things after the coroutineScope (starts at https://github.com/syncthing/syncthing-lite/pull/57/files/59ba2c324edb571db6399ae51ddf7ecae96830d0#diff-e654db538220d7523a566b9a386282edR84) are only executed when all tasks which are running in this scope have finished (or were canceled)